pkg/openapiconv/convert.go

```diff
@@ -0,0 +1,270 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapiconv
+
+import (
+	"strings"
+
+	klog "k8s.io/klog/v2"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+var OpenAPIV2DefPrefix = "#/definitions/"
+var OpenAPIV3DefPrefix = "#/components/schemas/"
+
+func ConvertV2ToV3(v2Spec *spec.Swagger) *spec3.OpenAPI {
+	v3Spec := &spec3.OpenAPI{
+		Version:      "3.0.0",
+		Info:         v2Spec.Info,
+		ExternalDocs: ConvertExternalDocumentation(v2Spec.ExternalDocs),
+		Paths:        ConvertPaths(v2Spec.Paths),
+		Components:   ConvertComponents(v2Spec.SecurityDefinitions, v2Spec.Definitions, v2Spec.Responses, v2Spec.Produces),
+	}
+
+	return v3Spec
+}
+
+func ConvertExternalDocumentation(v2ED *spec.ExternalDocumentation) *spec3.ExternalDocumentation {
+	if v2ED == nil {
+		return nil
+	}
+	return &spec3.ExternalDocumentation{
+		ExternalDocumentationProps: spec3.ExternalDocumentationProps{
+			Description: v2ED.Description,
+			URL:         v2ED.URL,
+		},
+	}
+}
+
+func ConvertComponents(v2SecurityDefinitions spec.SecurityDefinitions, v2Definitions spec.Definitions, v2Responses map[string]spec.Response, produces []string) *spec3.Components {
+	components := &spec3.Components{
+		Schemas:         make(map[string]*spec.Schema),
+		SecuritySchemes: make(spec3.SecuritySchemes),
+		Responses:       make(map[string]*spec3.Response),
+	}
+	for s, schema := range v2Definitions {
+		components.Schemas[s] = ConvertSchema(&schema)
+	}
+
+	for s, securityScheme := range v2SecurityDefinitions {
+		components.SecuritySchemes[s] = ConvertSecurityScheme(securityScheme)
+	}
+	for r, response := range v2Responses {
+		components.Responses[r] = ConvertResponse(&response, produces)
+	}
+
+	return components
+}
+
+func ConvertSchema(v2Schema *spec.Schema) *spec.Schema {
+	if v2Schema == nil {
+		return nil
+	}
+	var v3Schema spec.Schema
+	v3Schema = *v2Schema
+
+	if refString := v2Schema.Ref.String(); refString != "" {
+		if idx := strings.Index(refString, OpenAPIV2DefPrefix); idx != -1 {
+			v3Schema.Ref = spec.MustCreateRef(OpenAPIV3DefPrefix + refString[idx+len(OpenAPIV2DefPrefix):])
+		} else {
+			klog.Errorf("Error: Swagger V2 Ref %s does not contain #/definitions\n", refString)
+		}
+	}
+
+	for key, property := range v2Schema.Properties {
+		v3Schema.Properties[key] = *ConvertSchema(&property)
+	}
+	if v2Schema.Items != nil {
+		v3Schema.Items.Schema = ConvertSchema(v2Schema.Items.Schema)
+		if len(v2Schema.Items.Schemas) > 0 {
+			v3Schema.Items.Schemas = []spec.Schema{}
+			for _, s := range v2Schema.Items.Schemas {
+				v3Schema.Items.Schemas = append(v3Schema.Items.Schemas, *ConvertSchema(&s))
+			}
+		}
+	}
+
+	if v2Schema.AdditionalProperties != nil {
+		v3Schema.AdditionalProperties.Schema = ConvertSchema(v2Schema.AdditionalProperties.Schema)
```

Comment:
What is the test coverage for this function? I am not seeing where `v3Schema.AdditoinalProperties` is initialized.

Reply:
v3Schema is copied from v2Schema L86, so if it's not nil for v2, it's not nil for v3. But shouldn't it also be copied?